### PR TITLE
Fix fcpath

### LINKED ISSUE #95450: failed to find the FC disks starts with fc prefix
**What happened**:

Currently, we have logic here to look for the disk names with `-fc` but this fails when we have devices which are starting with `fc` https://github.com/kubernetes/kubernetes/blob/ee297b6f4d78daaf2f71c78a05203698581ca40a/pkg/volume/fc/fc_util.go#L63 

e.g: on ppc64le architecture, NPIV based fc disks are discovered as follows:

```shell
[root@dhcp-test-1 kubernetes]# ls -l /dev/disk/by-path/
total 0
lrwxrwxrwx. 1 root root  9 Oct  9 08:32 fc-0x5005076810213b32-lun-0 -> ../../sdf
lrwxrwxrwx. 1 root root 10 Oct  9 08:32 fc-0x5005076810213b32-lun-0-part1 -> ../../sdf1
lrwxrwxrwx. 1 root root 10 Oct  9 08:32 fc-0x5005076810213b32-lun-0-part2 -> ../../sdf2
lrwxrwxrwx. 1 root root  9 Oct  9 16:47 fc-0x5005076810213b32-lun-1 -> ../../sds
lrwxrwxrwx. 1 root root  9 Oct  9 16:35 fc-0x5005076810213b32-lun-2 -> ../../sdt
lrwxrwxrwx. 1 root root  9 Oct  9 14:37 fc-0x5005076810213b4f-lun-0 -> ../../sde
lrwxrwxrwx. 1 root root 10 Oct  9 08:32 fc-0x5005076810213b4f-lun-0-part1 -> ../../sde1
lrwxrwxrwx. 1 root root 10 Oct  9 08:32 fc-0x5005076810213b4f-lun-0-part2 -> ../../sde2
lrwxrwxrwx. 1 root root  9 Oct  9 16:35 fc-0x5005076810213b4f-lun-1 -> ../../sdq
lrwxrwxrwx. 1 root root  9 Oct  9 16:35 fc-0x5005076810213b4f-lun-2 -> ../../sdr
lrwxrwxrwx. 1 root root  9 Oct  9 08:32 fc-0x5005076810223b32-lun-0 -> ../../sdd
lrwxrwxrwx. 1 root root 10 Oct  9 08:32 fc-0x5005076810223b32-lun-0-part1 -> ../../sdd1
lrwxrwxrwx. 1 root root 10 Oct  9 08:32 fc-0x5005076810223b32-lun-0-part2 -> ../../sdd2
lrwxrwxrwx. 1 root root  9 Oct  9 16:35 fc-0x5005076810223b32-lun-1 -> ../../sdo
lrwxrwxrwx. 1 root root  9 Oct  9 16:35 fc-0x5005076810223b32-lun-2 -> ../../sdp
```

and here is the udev information for one of the disk:

```shell
[root@dhcp-test-1 kubernetes]# udevadm info -q all -n /dev/sds
P: /devices/vio/30000008/host2/rport-2:0-1/target2:0:1/2:0:1:1/block/sds
N: sds
S: disk/by-id/scsi-360050768108181da7800000000001b07
S: disk/by-id/scsi-SIBM_2145_02042060769eXX00
S: disk/by-id/wwn-0x60050768108181da7800000000001b07
S: disk/by-path/fc-0x5005076810213b32-lun-1
S: disk/by-path/fc-0xc050760b725f0568-0x5005076810213b32-lun-1
S: disk/by-uuid/314f3dcf-1b83-4571-9371-73d538dcaeb8
E: DEVLINKS=/dev/disk/by-id/wwn-0x60050768108181da7800000000001b07 /dev/disk/by-uuid/314f3dcf-1b83-4571-9371-73d538dcaeb8 /dev/disk/by-id/scsi-SIBM_2145_02042060769eXX00 /dev/disk/by-id/scsi-360050768108181da7800000000001b07 /dev/disk/by-path/fc-0xc050760b725f0568-0x5005076810213b32-lun-1 /dev/disk/by-path/fc-0x5005076810213b32-lun-1
E: DEVNAME=/dev/sds
E: DEVPATH=/devices/vio/30000008/host2/rport-2:0-1/target2:0:1/2:0:1:1/block/sds
E: DEVTYPE=disk
E: FC_INITIATOR_WWPN=0xc050760b725f0568
E: FC_TARGET_LUN=1
E: FC_TARGET_WWPN=0x5005076810213b32
E: ID_BUS=scsi
E: ID_FS_TYPE=ext4
E: ID_FS_USAGE=filesystem
E: ID_FS_UUID=314f3dcf-1b83-4571-9371-73d538dcaeb8
E: ID_FS_UUID_ENC=314f3dcf-1b83-4571-9371-73d538dcaeb8
E: ID_FS_VERSION=1.0
E: ID_MODEL=2145
E: ID_MODEL_ENC=2145\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20
E: ID_PATH=fc-0x5005076810213b32-lun-1
E: ID_PATH_TAG=fc-0x5005076810213b32-lun-1
E: ID_REVISION=0000
E: ID_SCSI=1
E: ID_SCSI_INQUIRY=1
E: ID_SCSI_SERIAL=02042060769eXX00
E: ID_SERIAL=360050768108181da7800000000001b07
E: ID_SERIAL_SHORT=60050768108181da7800000000001b07
E: ID_TARGET_PORT=0
E: ID_TYPE=disk
E: ID_VENDOR=IBM
E: ID_VENDOR_ENC=IBM\x20\x20\x20\x20\x20
E: ID_WWN=0x60050768108181da
E: ID_WWN_VENDOR_EXTENSION=0x7800000000001b07
E: ID_WWN_WITH_EXTENSION=0x60050768108181da7800000000001b07
E: MAJOR=65
E: MINOR=32
E: SCSI_IDENT_LUN_NAA_REGEXT=60050768108181da7800000000001b07
E: SCSI_IDENT_PORT_RELATIVE=2180
E: SCSI_IDENT_PORT_TARGET_PORT_GROUP=0x0
E: SCSI_IDENT_SERIAL=02042060769eXX00
E: SCSI_MODEL=2145
E: SCSI_MODEL_ENC=2145\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20
E: SCSI_REVISION=0000
E: SCSI_TPGS=1
E: SCSI_TYPE=disk
E: SCSI_VENDOR=IBM
E: SCSI_VENDOR_ENC=IBM\x20\x20\x20\x20\x20
E: SUBSYSTEM=block
E: TAGS=:systemd:
E: USEC_INITIALIZED=29384079103

[root@dhcp-test-1 kubernetes]#
```

Because of the existing logic k8s fails to detect the disk because it does not start with `-fc-0x` instead starts with `fc-0x`, so pod creation fails with the following error:

```shell
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Normal   Scheduled               36m                default-scheduler        Successfully assigned default/mkumatag-fc-pod to dhcp-test-1
  Normal   SuccessfulAttachVolume  36m                attachdetach-controller  AttachVolume.Attach succeeded for volume "fc-vol"
  Warning  FailedMount             33m                kubelet                  Unable to attach or mount volumes: unmounted volumes=[fc-vol], unattached volumes=[default-token-k6pl4 fc-vol]: timed out waiting for the condition
  Warning  FailedMount             33m (x5 over 35m)  kubelet                  MountVolume.WaitForAttach failed for volume "fc-vol" : no fc disk found
```

**What you expected to happen**:
FC disk should get attached properly to the pod

**How to reproduce it (as minimally and precisely as possible)**:
sample pod yaml on a machine with path names starts with `fc`
```
apiVersion: v1
kind: Pod
metadata:
 name: mkumatag-fc-pod
spec:
 containers:
   - image: ubuntu:latest
     command: ["/bin/sh", "-ec", "while :; do echo '.'; sleep 5 ; done"]
     name: fc
     volumeMounts:
       - name: fc-vol
         mountPath: /mnt/fc
 volumes:
   - name: fc-vol
     fc:
       targetWWNs:
         - 5005076810213b32
       lun: 1
       fsType: ext4
```

**Anything else we need to know?**:

**Environment**:
`N/A`

### PR DESCRIPTION
**Pull request by mkumatag**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #95450

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/volume/fc/fc_util.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func (handler *osIOHandler) WriteFile(filename string, data []byte, perm os.File
 
 // given a wwn and lun, find the device and associated devicemapper parent
 func findDisk(wwn, lun string, io ioHandler, deviceUtil volumeutil.DeviceUtil) (string, string) {
-	fcPath := "-fc-0x" + wwn + "-lun-" + lun
+	fcPath := "fc-0x" + wwn + "-lun-" + lun
```
Comment (saad-ali) [inline pkg/volume/fc/fc_util.go:63]: Can you please add a corresponding unit test in fc_utils_test.go

**Review by saad-ali — COMMENTED** [review]
I am not a FC expert, but the change looks reasonable to me.
/approve
/hold

@rootfs can you PTAL

File: pkg/volume/fc/fc_util.go (hunk 1 above)
Comment (mkumatag) [reply pkg/volume/fc/fc_util.go:63]: @saad-ali done, added the UT for the new scenario, and for the missing negative case as well.

**Review by saad-ali — COMMENTED** [review]
/lgtm

Want to give @rootfs a chance to look at this as well. Will let him remove hold.
If he is not able to take a look, feel free to ping me again.

**Comment by rootfs** [comment]
I don't have a FC setup to verify, but this could be due to a difference in udev rules. @mkimuram @mtanino can you check?

File: pkg/volume/fc/fc_util.go (hunk 1 above)
Comment (rootfs) [inline pkg/volume/fc/fc_util.go:63]: This regex is too loose. Unless there are other naming conventions, we only have `/dev/disk/by-path/fc-0x` and `/dev/disk/by-path/pci-xxxx-fc-0x`. The regex should just filter these cases.

Comment (mkumatag) [reply pkg/volume/fc/fc_util.go]: @rootfs added the regexp for the above cases, can you ptal at the latest changes?

**Comment by mkimuram** [comment]
The name depends on udev rules, so it makes sense to add expressions that match to other naming conventions.
(I haven't tested with the actual environment, but as https://github.com/hreinecke/sg3_utils/blob/master/scripts/59-fc-wwpn-id.rules shows, distributions that uses sg3_utils seems to name with the convention in this PR.)

So, the overall change looks good to me. 

I would like to confirm two things:

1.  Should we care the name that contains "-fc-0x" but doesn't start with "pci-"?
It was allowed by the original code, but will be denied by the new code. 
I'm not sure that there are such naming conventions used in a real environment, but would like to confirm that it is OK, just for sure.

2. Should we need to care a distribution that has both "pci-xxx-fc-0xYYYY" and " fc-0xYYYY"?
In such a distribution, the path found first in the loop will be used. 
I believe that this won't make any issue for the device name is resolved from the symlink and the symlink should point to the same device, but I would like to make sure that it won't be an issue.

**Comment by mkimuram** [comment]
@rootfs 

> 1. Should we care the name that contains "-fc-0x" but doesn't start with "pci-"?

I read through the discussion again and your comment "Unless there are other naming conventions, we only have /dev/disk/by-path/fc-0x and /dev/disk/by-path/pci-xxxx-fc-0x. The regex should just filter these cases." seems to say that we don't need to care about this case. I'm OK with it, if it is practically good.

Then, could you just double check "2. Should we need to care a distribution that has both "pci-xxx-fc-0xYYYY" and " fc-0xYYYY"?"?

**Comment by rootfs** [comment]
thanks @mkimuram 
for the 2nd case when both paths exist (which should be considered poor configuration), then the first match is returned.

**Pull request merged**